azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +73,176 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
         else:
             logger.error(msg)
         if send_event:
-            add_event(op=op, message=msg, is_success=is_success)
+            add_event(op=op, message=msg, is_success=is_success, log_event=False)
 
     @staticmethod
-    def is_policy_enforcement_enabled():
+    def __get_policy_enforcement_enabled():
         """
-        Check whether user has opted into policy enforcement feature.
-        Caller function should check this before performing any operations.
+        Policy will be enabled if (1) policy file exists at _CUSTOM_POLICY_PATH and (2) the conf flag "Debug.EnableExtensionPolicy" is true.
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
         # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        policy_file_exists = os.path.exists(_CUSTOM_POLICY_PATH)
+        return conf.get_extension_policy_enabled() and policy_file_exists
+
+    @property
+    def policy_enforcement_enabled(self):
+        return self._policy_enforcement_enabled
+
+    def __get_policy(self):
+        """
+        Load policy JSON object from policy file (CUSTOM_POLICY_PATH), return as a dict.
+
+        Note that we should only call this function after validating that CUSTOM_POLICY_PATH exists (this is currently
+        done in __init__).
+        """
+        # TODO: Add schema validation for custom policy file and raise relevant error message to user.
+        with open(_CUSTOM_POLICY_PATH, 'r') as f:        # Open file in read-only mode.
+            self._log_policy_event("Custom policy file found at {0}. Using custom policy.".format(_CUSTOM_POLICY_PATH))
+            try:
+                custom_policy = json.load(f)
```

Comment:
We read the policy on a per goal state basis. If the customer does change the file during goal state processing, we don't necessarily want to support that so it would make sense to raise an error.

Reply:
If user is modifying policy file while the agent is processing extensions, what message will be surfaced to the user? Will it be useful/clear?

Reply:
The policy file is only read once during extension processing. If the user successfully modifies the file to a valid JSON before json.load is called, no error message is raised and we apply the modified version of the policy. 

If the file is actively being modified and incomplete/invalid when json.load is called, JSONDecodeError will be thrown, we catch and re-raise it as an InvalidPolicyError with an error message like "policy file does not conform to valid json syntax". This might be a case where it makes sense to dump the policy file in the error message.

Reply:
I think we should log on read. If we do that, then we won't need to add it to the error message

Reply:
Oh I see you meant add it to error message for user to see. That could be helpful, but as you pointed out in the other thread may be overkill. For now, maybe add it to local log and not status?